### DISCUSSION TITLE
How to make a layour from state

### DISCUSSION THREAD

**Discussion by MrNemo64** [Textualize/textual#5178] [Q&A] [answered]
I've been looking at the docks about layouts, specifically these two links

https://textual.textualize.io/how-to/design-a-layout/

https://textual.textualize.io/guide/layout/

Now I want to make a layourt where I have a component on one half and two stacked on the other half but when making the layout I keep getting an exception `AttributeError: 'View' object has no attribute 'logger'`
```py
class View(App):
    logger: LogView
    taxi: Taxi

    async def on_mount(self):
        self.logger = LogView()
        self.taxi = Taxi(self.logger)
        self.taxi.iniciar()
    
    def compose(self) -> ComposeResult:
        yield Vertical(
            Horizontal(
                self.logger, # Extends ScrollView
                self.taxi.view_sensores() # Extends ScrollView
            )
        )

    async def on_key(self, event):
        if event.key == "q":
            await self.action_quit()

    def get_logger(self):
        return self.logger
```
How am I supposed to make a layout from the state of my app?

**Comment by TomJGooding** [comment]
The problem is that your class variables haven't been initialized, because `on_mount` is called *after* `compose`.

Could you explain what you're trying to achieve, as I'm afraid I'm not sure I follow?

**Reply by MrNemo64** [reply, marked as answer]
I was trying to hold onto a reference of the components to be able to use their methods. After looking thru the examples folder I was able to figure out that the way to do this is using the `query_one` method so I did some refactoring now it works!
```py
class View(App):
    async def on_mount(self):
        logger = self.query_one(LogView)
        mapa = self.query_one(MapView)
        taxi = Taxi(logger, self.query_one(ViewSensores), mapa)
        await taxi.iniciar()
    
    def compose(self) -> ComposeResult:
        yield Horizontal(
            VerticalScroll(MapView()),
            Vertical(
                VerticalScroll(LogView()), 
                ViewSensores()
            )
        )

    async def on_key(self, event):
        if event.key == "q":
            await self.action_quit()
            sys.exit(0)
```
